이미지 리사이징을 통한 메모리 사용량 개선

Hansolkkim opened the pull request:
close #79 

- [x] Image Resizing을 통한 메모리 사용량 개선

(트러블 슈팅 내용을 일단 여기에 정리해놓겠습니다!)

### 이미지와 메모리 최적화에 대한 고민

* 기존에는 아래와 같은 로직으로 UIImageView에 이미지를 넣었습니다.

```swift
func configure(with photo: Photo) {
    let image = UIImage(data: photo.imageData)

    thumbnailImageView.image = image
}
```

하지만 `Data` 타입을 이용해 UIImage를 렌더링하는 초기화 생성자가 스레드에 상당한 부담을 준다는 사실을 확인(여러 이미지를 보여주는 AlbumView에서 사용 메모리가 120MB 까지 올라가는 현상이 있었음)하여, 다른 스레드를 사용해 메인스레드에서의 부담을 줄여주어 메모리 부담을 개선할 수 있었습니다.

```swift
func configure(with photo: Photo) {
    let serialQueue = DispatchQueue(label: "Serial_Queue")

    serialQueue.async {
        let image = UIImage(data: photo.imageData)

        DispatchQueue.main.async {
            thumbnailImageView.image = image
        }
    }
}
```



* 하지만 추후 이미지에 Text를 얹어 보여주는 기능을 추가함에 따라 메모리 부담이 다시 늘어나는 문제가 있었습니다. 이미지에 Text를 얹어 보여주는 기능은 아래의 메소드를 사용했습니다.

```swift
extension UIImage {
    func overlayText(of photoData: Photo) -> UIImage? {
        let text = DateConverter.convertToDateString(date: photo.date)
        let textColor = UIColor(hexRGB: photo.color) ?? .clear
        let fontSize: CGFloat = 0.1 * self.size.height
        let imageSize = CGSize(width: size.width, height: size.height)
        let textFont = UIFont(name: photo.style.name, size: fontSize) ?? UIFont(name: Style.none.rawValue, size: fontSize) ?? .systemFont(ofSize: 0)

        let attributedString = NSMutableAttributedString(string: text)
        attributedString.addAttributes(
             [NSAttributedString.Key.font: textFont,
                NSAttributedString.Key.foregroundColor: textColor],
            range: (text as NSString).range(of: text)
        )
    
        UIGraphicBeginImageContextWithOptions(imageSize, false, 1.0)
        self.draw(in: CGRect(origin: .zero, size: imageSize)

        let label = UILabel()
        label.numberOfLines = 0
        label.textAlginment = .center
        label.attributedText = attributedString
        label.drawText(in: CGRect(origin: .zero, size: imageSize))

        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return newImage
}
```

해당 메소드에서 UILabel을 사용하고 있기 때문에 SerialQueue로 해당 작업을 보낼 수 없어 MainQueue에서 작업을 진행하도록 했는데, UILabel을 사용하지 않고 `NSString.draw(in:)`메소드를 사용할 수 있어 해당 방법을 사용했으나 메모리 사용량에 개선이 많지 않았습니다.

```swift
extension UIImage {
    func overlayText(of photoData: Photo) -> UIImage? {
        let text = DateConverter.convertToDateString(date: photo.date)
        let textColor = UIColor(hexRGB: photo.color) ?? .clear
        let fontSize: CGFloat = 0.1 * self.size.height
        let imageSize = CGSize(width: size.width, height: size.height)
        let textFont = UIFont(name: photo.style.name, size: fontSize) ?? UIFont(name: Style.none.rawValue, size: fontSize) ?? .systemFont(ofSize: 0)
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center

        let attributedString = NSMutableAttributedString(string: text)
        attributedString.addAttributes(
             [NSAttributedString.Key.font: textFont,
                NSAttributedString.Key.foregroundColor: textColor,
                NSAttributedString.Key.paragraphStyle: paragraph]
            range: (text as NSString).range(of: text)
        )
    
        UIGraphicBeginImageContextWithOptions(imageSize, false, 1.0)
        self.draw(in: CGRect(origin: .zero, size: imageSize)

        attributedString.draw(in: CGRect(origin: .zero, size: imageSize)) // <--

        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return newImage
}
```

따라서 추후 Photo.Style을 현재 사용하고 있는 글씨체(UIFont) 변경의 기능에서 여러 모양을 가지는 UIView를 추가할 예정이고, 메모리 사용량 개선에 큰 의미없는 부분 때문에 억지로 SerialQueue에서 처리 가능하도록 할 필요가 없다고 생각하여, UILabel을 사용하는 코드로 계속 사용하도록 했습니다.

* 또한 위의 코드에서 이미지를 렌더링하는 때에 사용한 `UIGraphicsBeginImageContextWithOptions`의 경우 SRGB format의 색영역 사용을 강제한다는 문제가 있어, 메모리 개선을 위해 iOS 10부터 사용할 수 있는 `UIGraphicsImageRenderer`를 사용해 컴파일러가 적절한 색영역 사용을 선택할 수 있도록 하여 메모리 사용을 개선할 수 있다고 판단해 아래의 코드를 사용해보았습니다.

```swift
extension UIImage {
    func overlayText(of photoData: Photo) -> UIImage? {
        let text = DateConverter.convertToDateString(date: photo.date)
        let textColor = UIColor(hexRGB: photo.color) ?? .clear
        let fontSize: CGFloat = 0.1 * self.size.height
        let imageSize = CGSize(width: size.width, height: size.height)
        let textFont = UIFont(name: photo.style.name, size: fontSize) ?? UIFont(name: Style.none.rawValue, size: fontSize) ?? .systemFont(ofSize: 0)
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center

        let attributedString = NSMutableAttributedString(string: text)
        attributedString.addAttributes(
             [NSAttributedString.Key.font: textFont,
                NSAttributedString.Key.foregroundColor: textColor,
                NSAttributedString.Key.paragraphStyle: paragraph],
            range: (text as NSString).range(of: text)
    )
    
        let renderer = UIGraphicsImageRenderer(size: imageSize)
        return renderer.image { context in
                self.draw(in: CGRect(origin: .zero, size: imageSize))

                let label = UILabel()
                label.numberOfLines = 0
                label.textAlignment = .center
                label.attributedText = attributedString
                label.drawText(in: CGRect(origin: .zero, size: imageSize))
    }
}
```

하지만, 여러 색이 들어있는 UIImage도 렌더링하기 때문에 sRGB 색영역을 사용할텐데 컴파일러가 사용할 색 영역을 선택하는 과정에서의 메모리 부하가 발생하여 기존보다 메모리 사용량이 크게 늘어나는 문제가 발생한다고 판단하여 해당 코드도 사용하지 않도록 했습니다.

| `UIGraphicsBeginImageContextWithOptions` 사용시              | `UIGraphicsImageRenderer` 사용시                             |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![SS2022-10-20PM09.54.41](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202210202210713.jpg) | ![SS2022-10-20PM10.08.07](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202210202210834.jpg) |



* 마지막으로 생각한 개선의 여지는 UIImage를 만들 때 `UIImage(data:)` 초기화 생성자가 아닌, Data에 있는 이미지 정보를 CGImage 단계에서 적절한 크기의 이미지로 Downsampling 한 후에 UIImage에 그려주는 방법입니다. 이미지를 Downsampling 해주는 코드는 아래의 코드를 사용했습니다.

```swift
extension Data {
    func toDownsampledImage(pointSize: CGSize, scale: CGFloat) -> UIImage? {
        let imageSourceOptions = [kCGImageSourceShouldCache: false] as CFDictionary
        guard let imageSource = CGImageSourceCreateWithData(self as CFData, imageSourceOptions) else {
            return nil
        }
        let maxDimensionInPixels = Swift.max(pointSize.width, pointSize.height) * scale
        let downsampleOptions = [
            kCGImageSourceCreateThumbnailFromImageAlways: true,
            kCGImageSourceShouldCacheImmediately: true,
            kCGImageSourceCreateThumbnailWithTransform: true,
            kCGImageSourceThumbnailMaxPixelSize: maxDimensionInPixels
        ] as CFDictionary

        if let downsampledImage = CGImageSourceCreateThumbnailAtIndex(imageSource, 0, downsampleOptions) {
            return UIImage(cgImage: downsampledImage)
        }

        return nil
    }
}
```

1. `kCGImageSourceCreateThumbnailFromImageAlways`

    - 원본 이미지 소스 파일에 Thumbnail이 있는지, 원본 이미지에서 Thumbnail을 새로 만들어야하는지를 설정하는 옵션입니다.

    - 해당 값을 true로 설정하면 `kCGImageSourceThumbnailMaxPixelSize`에 지정된 제한에 따라 full image로부터 Thumbnail을 만들고, `kCGImageSourceThumbnailMaxPixelSize`를 지정하지 않으면 최대 원본 이미지 크기만큼의 Thumbnail을 만듭니다.

2. `kCGImageSourceShouldCacheImmediately`

    * Thumbnail 생성을 요청 하는 시점에 Core Graphics에게 이미지 디코딩을 시작하라는 알림을 주는 옵션입니다.
    * false로 설정할 경우(default), 이미지 데이터 객체를 생성할 뿐이니 이미지 소스에 대한 참조만 생성하고, 곧바로 이미지 디코딩을 시작하지 말라는 의미의 옵션입니다.

3. `kCGImageSourceThumbnailMaxPixelSize`

    * Thumbnail의 최대 픽셀을 지정하는 옵션입니다.

4. `kCGImageSourceCreateThumbnailWithTransform`

    - 원본 이미지의 방향 및 종횡비에 따라 Thumbnail을 회전 및 스케일링해야하는지를 설정하는 옵션입니다.

    - 해당 옵션을 true로 설정함으로써 Core Graphics에게 원본 이미지와 Downsampling된 이미지의 orientation이 같도록 설정할 수 있습니다.

위의 코드를 사용한 Downsampling으로 이미지로 인한 메모리 사용량을 개선할 수 있었습니다. 또한 Downsampling된 이미지를 사용할 경우 메모리가 순간적으로 큰 값으로 튀는 구간도 어느정도 제어할 수 있었기 때문에 그로 인한 개선도 할 수 있었습니다.

| Downsampling 이전                                            | Downsampling 이후                                            |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| ![SS2022-10-20PM09.54.41](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202210202201675.jpg) | ![SS2022-10-20PM09.48.39](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202210202252556.jpg) |

Hansolkkim committed:
refactor: [#79] UIImage로 인한 메모리 사용량 개선

- downsampling하는 로직을 통해, Data가 UIImage로 변환되기 전에 View의 사이즈에 맞는 scale, pixel size를 가지도록 downsampling한 후에 변환하도록 하여 메모리 사용량 개선

Hansolkkim committed:
refactor: [#79] Cell Image의 2중 스케일 다운 제거 및 PhotoEditView의 imageView에도 downsampling image 로직 적용

Hansolkkim commented on rabit/rabit/Utilities/Data + Utils.swift:
```diff
@@ -0,0 +1,29 @@
+            return nil
+        }
+        let maxDimensionInPixels = Swift.max(pointSize.width, pointSize.height) * scale
+        let downsampleOptions = [
+            kCGImageSourceCreateThumbnailFromImageAlways: true,
+            kCGImageSourceShouldCacheImmediately: true,
+            kCGImageSourceCreateThumbnailWithTransform: true,
+            kCGImageSourceThumbnailMaxPixelSize: maxDimensionInPixels
+        ] as CFDictionary
```
옵션이 다양한데, 일단 보편적으로 사용하는 옵션들만 사용했습니다. 
사용한 옵션의 의미는 PR 내용에 작성했습니다!

junu0516 replied:
본문 읽어봤어요 ㅎㅎ 설명 감사합니다!

Hansolkkim commented on rabit/rabit/Album/StyleSelect/Views/StyleSelectCell.swift:
```diff
@@ -58,8 +58,13 @@ final class StyleSelectCell: UICollectionViewCell {
+        )
+
         serialQueue.async {
-            let image = UIImage(data: photo.imageData)
+            let image = photo.imageData.toDownsampledImage(pointSize: imageSize, scale: 2.0)
 
```
현재는 화면에 최대로 이미지가 많이 보여봤자 6개 정도인데, 아이패드처럼 큰 화면에서는 한 화면에 더 많은 이미지가 보여질 수도 있고 그 때 이 코드로 인해 스레드가 너무 많이 생겨나는 문제도 발생할 것 같습니다.
따라서 큐를 하나 정도로 제한하여, VC가 Cell 마다 동일한 Queue를 사용하게 하면 될 거 같은데, 일단은 아이패드는 상정하지 않기 때문에 해당 코드를 유지했습니다!

junu0516 replied:
이건 결국 글로벌 큐 사용하는 부분으로 대체된거라 해결된 걸로 간주해도 괜찮죠!!?

Hansolkkim replied:
넵 어느 정도 해결된거 같아요!ㅎㅎ

Hansolkkim commented on rabit/rabit/Utilities/UIImage + Utils.swift:
```diff
@@ -39,11 +39,11 @@ extension UIImage {
-        let textFont = UIFont(name: photo.style.name, size: fontSize) ?? UIFont.systemFont(ofSize: 100)
-        
+        let textColor = UIColor(hexRGB: photo.color) ?? .clear
+        let fontSize: CGFloat = 0.1 * size.height
+        let imageSize = CGSize(width: size.width, height: size.height)
+        let textFont = UIFont(name: photo.style.name, size: fontSize) ?? .systemFont(ofSize: 0.1 * size.height)
```
어제 제드와 함께 작성했던 코드입니다ㅋㅋ
Text의 위치, 크기가 AlbumView, PhotoEditView, StyleSelectView에서 모두 같도록 하기 위해 여러 방법을 고심하다가 UIFont의 크기를 변경하면 세 화면에서 모두 상당히 똑같은 위치에 위치하게 된다는것을 확인하여 이처럼 코드를 작성했습니다!

junu0516 replied:
크... 장인정신 다시 한 번 리스펙하고 갑니다 ㅋㅋㅋ

Hansolkkim commented on rabit/rabit/Utilities/UIImage + Utils.swift:
```diff
@@ -37,11 +37,11 @@ extension UIImage {
         return UIImage(cgImage: croppedImage, scale: 1.0, orientation: self.imageOrientation)
     }
     
-    func overlayText(of photo: Photo, scale: CGFloat = UIScreen.main.scale) -> UIImage? {
```
기존에 Downsampling을 하지 않을 때는 이렇게 Text를 얹을 때 사진의 scale을 줄여 해상도를 낮춰주었었는데,
Downsampling 과정이 추가됨으로 인해 scale을 줄이게 된다면 2중으로 이미지의 해상도가 낮아지는 문제가 있어 해당 메소드에서는 scale을 건드리지 않기로 했습니다.

Hansolkkim commented on rabit/rabit/Utilities/Data + Utils.swift:
```diff
@@ -0,0 +1,29 @@
+import UIKit
+
+extension Data {
+    func toDownsampledImage(pointSize: CGSize, scale: CGFloat = UIScreen.main.scale) -> UIImage? {
```
이 메소드는 UIImage가 들어갈 View의 Size를 넘겨받아 해당 Size와 적절한 scale에 맞는 픽셀수를 정한 후, UIImage의 데이터에 해당하는 Data를 앞서 정한 픽셀 수로 제한하여 썸네일으로 생성해주는 메소드입니다. 저도 처음보는 문법이 많네요....ㅋㅋ

Hansolkkim committed:
refactor: [#79] Cell configure에 사용하는 Queue 변경

- Serial Queue에 대해 잘못 생각하고 있어, 굳이 매번 새로 Queue를 생성하지 않고 글로벌 큐로 작업을 보내도록 코드 수정

Hansolkkim committed:
chore: [#79] 커밋 오류 수정

Hansolkkim commented on rabit/rabit/Album/Views/AlbumCell.swift:
```diff
@@ -24,10 +24,14 @@ final class AlbumCell: UICollectionViewCell {
+            height: bounds.width - 20
+        )
+
+        DispatchQueue.global().async {
```
Serial Queue를 다시 한 번 잘 못 이해했네요..ㅋㅋㅋ
저는 클로저 내에 있는 image downsampling 이후에 imageView에 image를 넣는 작업을 순서대로 하길 원해서 Serial Queue를 사용했는데 제가 원하는 대로 동작하는게 아니었어요..ㅋㅋㅋ
클로저 내에 있는 코드는 어짜피 순서대로 동작하고, Serial Queue가 받는 클로저들이 순차적으로 쌓이는 건데 오해를 하고 있었네요..

굳이 매 Cell을 그릴 때마다 Queue를 새로 생성하지 않도록 글로벌큐를 그냥 사용하도록 했습니다!

junu0516 commented:
@Hansolkkim 
노션 글이랑 같이 PR 쭉 봤어요 ㅋㅋ
학습 조금만 더 하고.. PR을 다시 봐야 이해할 수 있을듯해서 ㅠㅠ 주말안에 코멘트 달 수 있도록 해볼께요!!
수고 정말 많으셨어요 ㅎㅎ

junu0516 commented on rabit/rabit/Album/PhotoEdit/PhotoEditViewController.swift:
```diff
@@ -102,9 +102,16 @@ private extension PhotoEditViewController {
     func bind() {
         guard let viewModel = viewModel else { return }
 
+        let imageSize = CGSize(
+            width: view.bounds.width,
+            height: view.bounds.width
+        )
```
이미지의 변의 길이를 나타내는 지역변수를 하나 추가해도 괜찮을 것 같아요!
```suggestion
        let imageLength = view.bounds.width
        let imageSize = CGSize(
            width: imageLength,
            height: imageLength
        )
```

junu0516 commented on rabit/rabit/Utilities/Data + Utils.swift:
```diff
@@ -0,0 +1,29 @@
+            kCGImageSourceThumbnailMaxPixelSize: maxDimensionInPixels
+        ] as CFDictionary
+
+        if let downsampledImage = CGImageSourceCreateThumbnailAtIndex(
+            imageSource,
+            0,
+            downsampleOptions
+        ) {
+            return UIImage(cgImage: downsampledImage)
+        }
+
+        return nil
```
이건 사견입니다!

Data에 대한 extension이라서.. 다운샘플링된 cgImage 데이터 자체만 옵셔널 타입으로 리턴해도 괜찮을듯 해요
(만들어주신 Data extension 메소드의 리턴타입을 Data?로 두면 return nil 같은 것들을 제거할 수 있을 것 같습니다.)

이후에 UIImage로의 변환은 외부에서 다운샘플링된 cgImage 데이터로 자유롭게 할 수 있도록 하면 외부에서는 compactMap 사용하는 식으로 좀 더 직관적으로 사용할 수 있을 것 같아요!

```swift
extensino Data {
    func toDownsampledImage(pointSize: CGSize, scale: CGFloat = UIScreen.main.scale) -> Data? {
        //cgImage 자체를 옵셔널로 리턴
    }
}
// 외부에서는 cgImage를 받아서 자유롭게 사용
func bind() {
    someObservableData
                .compactMap { $0.toDownsampledImage }
                .map { UIImage(data:$0) }
}

Hansolkkim replied:
음 안그래도 Data타입 extension에 UIKit을 import해서 불편한 생각이 있었는데, 이렇게 하면 괜찮을 것 같네요ㅎㅎ
감사합니다!

junu0516 commented on rabit/rabit/Album/Views/AlbumCell.swift:
```diff
@@ -24,15 +24,14 @@ final class AlbumCell: UICollectionViewCell {
 
-        serialQueue.async {
             let image = photo.imageData.toDownsampledImage(pointSize: imageSize, scale: 2.0)
+        DispatchQueue.global().async {
```
qos를 default보다 좀 더 높여도 괜찮을 것 같아요

junu0516 reviewed (approved):
수고하셨어요 ㅎㅎ 덕분에 저도 하나 더 배워갑니다.
한 가지 아래 코드 보고 궁금한게 있어서 질문 남길께요!
(나머지는 제안사항 보시고 크게 수정하거나 더 이상 반영할거 없으시면 머지해도 될 것 같아요!)

```swift
extension UIImage {
    func overlayText(of photoData: Photo) -> UIImage? {
        let text = DateConverter.convertToDateString(date: photo.date)
        let textColor = UIColor(hexRGB: photo.color) ?? .clear
        let fontSize: CGFloat = 0.1 * self.size.height
        let imageSize = CGSize(width: size.width, height: size.height)
        let textFont = UIFont(name: photo.style.name, size: fontSize) ?? UIFont(name: Style.none.rawValue, size: fontSize) ?? .systemFont(ofSize: 0)
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center

        let attributedString = NSMutableAttributedString(string: text)
        attributedString.addAttributes(
             [NSAttributedString.Key.font: textFont,
                NSAttributedString.Key.foregroundColor: textColor,
                NSAttributedString.Key.paragraphStyle: paragraph],
            range: (text as NSString).range(of: text)
    )
    
        let renderer = UIGraphicsImageRenderer(size: imageSize)
        return renderer.image { context in
                self.draw(in: CGRect(origin: .zero, size: imageSize))

                let label = UILabel()
                label.numberOfLines = 0
                label.textAlignment = .center
                label.attributedText = attributedString
                label.drawText(in: CGRect(origin: .zero, size: imageSize))
    }
}
```
위의 코드에서 `UIGraphicsImageRenderer` 을 사용할 때, 이미지 랜더링 시에 호출하는 클로저 안에서 UILabel을 사용하는데 이 부분을 CATextLayer을 사용하는 식으로 좀 더 가볍게 처리할 수 있지 않을까 하는 생각이 들었습니다.
지난번에 CALayer에 대해 이야기하면서 잠깐 언급했는데, 다시 한 번 CATextLayer에 대해서 찾아보니까 UIView의 인스턴스를 만드는 것보다 더 가벼워서 괜찮을 것 같더라구요!

제가 우선적으로 참고한 [링크](https://velog.io/@yongchul/iOSCALayer%EC%9D%98-%EA%B8%B0%EB%B3%B8-%EA%B0%9C%EB%85%90) 를 보고 판단했습니다!

조금 요약해보면 정적인 상황이라면 UILabel을 사용하는게 괜찮지만, 애니메이션이나 동적으로 바뀔 여지가 많은 경우에는 CATextLayer도 나쁘지 않다는 것 같아요.
(이건 만났을 때 한번 더 여쭤볼께요 ㅋㅋ 저도 레퍼런스 좀만 더 찾아봐야겠습니다)

Hansolkkim committed:
chore: [#79] 지역 변수 추가

Co-authored-by: Junu Lee(Jed) <junu0516@kakao.com>

Hansolkkim committed:
refactor: [#79] PR 의견 반영

- `return nil` 과 같은 불필요한 코드 제거와 좀 더 직관적인 메소드 사용을 위해 Data 타입의 toDownsampledImage 이름 변경 및 반환타입 변경
- UI의 부드러운 변경을 위해 AlbumView, StyleSelectView의 CollectionView에 들어갈 이미지를 생성해주는 큐의 qos를 높은 qos로 변경
